Sources/Plasma/PubUtilLib/plPipeline/pl3DPipeline.h

```diff
@@ -0,0 +1,770 @@
+    hsGDeviceRef*                       fCurrRenderTargetRef;
+
+    uint32_t                            fOrigWidth;
+    uint32_t                            fOrigHeight;
+    uint32_t                            fColorDepth;
+
+    uint32_t                            fInSceneDepth;
+    double                              fTime;      // World time.
+    uint32_t                            fFrame;     // inc'd every time the camera moves.
+    uint32_t                            fRenderCnt; // inc'd every begin scene.
+
+    bool                                fVSync;
+
+
+public:
+    pl3DPipeline(const hsG3DDeviceModeRecord* devModeRec);
+    virtual ~pl3DPipeline();
+
+    CLASSNAME_REGISTER(pl3DPipeline);
+    GETINTERFACE_ANY(pl3DPipeline, plPipeline);
+
+    size_t GetViewStackSize() const { return fViewStack.size(); }
+
+
+    /*** VIRTUAL METHODS ***/
+    //virtual bool PreRender(plDrawable* drawable, hsTArray<int16_t>& visList, plVisMgr* visMgr=nullptr) = 0;
+    //virtual bool PrepForRender(plDrawable* drawable, hsTArray<int16_t>& visList, plVisMgr* visMgr=nullptr) = 0;
+
+
+    /**
+     * The normal way to render a subset of a drawable.
+     *
+     * This assumes that PreRender and PrepForRender have already been called.
+     *
+     * Note that PreRender and PrepForRender are called once per drawable per
+     * render with a visList containing all of the spans which will be
+     * rendered, but Render itself may be called with multiple visList subsets
+     * which union to the visList passed into PreRender/PrepForRender. This
+     * happens when drawing sorted spans, because some spans from drawable B
+     * may be in the middle of the spans of drawable A, so the sequence would
+     * be:
+     *    PreRender(A, ATotalVisList);
+     *    PreRender(B, BTotalVisList);
+     *    PrepForRender(A, ATotalVisList);
+     *    PrepForRender(B, BTotalVisList);
+     *    Render(A, AFarHalfVisList);
+     *    Render(B, BTotalVisList);
+     *    Render(A, ANearHalfVisList);
+     *
+     * See plPageTreeMgr, which handles all this.
+     */
+    virtual void Render(plDrawable* d, const hsTArray<int16_t>& visList);
+
+
+    /**
+     * Convenience function for a drawable that needs to get drawn outside of
+     * the normal scene graph render (i.e. something not managed by the
+     * plPageTreeMgr).
+     *
+     * Not nearly as efficient, so only useful as a special case.
+     */
+    virtual void Draw(plDrawable* d);
+
+
+    //virtual plTextFont* MakeTextFont(char* face, uint16_t size) = 0;
+    //virtual void CheckVertexBufferRef(plGBufferGroup* owner, uint32_t idx) = 0;
+    //virtual void CheckIndexBufferRef(plGBufferGroup* owner, uint32_t idx) = 0;
+    //virtual bool OpenAccess(plAccessSpan& dst, plDrawableSpans* d, const plVertexSpan* span, bool readOnly) = 0;
+    //virtual bool CloseAccess(plAccessSpan& acc) = 0;
+    //virtual void CheckTextureRef(plLayerInterface* lay) = 0;
```

Comment:
These are all the virtual methods from plPipeline. I added them as comments so that it's easy to see which functions still need to be looked at for possible promotion up from plDXPipeline.